core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -286,6 +286,7 @@ class DAGScheduler(
       case None =>
         // We are going to register ancestor shuffle dependencies
         getAncestorShuffleDependencies(shuffleDep.rdd).foreach { dep =>
+          assert(!shuffleToMapStage.get(dep.shuffleId).isDefined)
```

Comment:
Overwriting entries in `shuffleToMapStage` definitely isn't an outright error -- we should still get correct results after the overwrite; so we shouldn't be adding this new assertion to change an evaluation path that was generating correct results into an error condition.

There are other open efforts to clean up this creation of additional Stages that will be skipped instead of just re-using the prior-executed Stage more cleanly.  That issue is orthogonal to the generation of a topological ordering of the dependencies, so I'd prefer to handle it outside of this PR.

cc @squito